.circleci/config.yml

```diff
@@ -1,22 +1,150 @@
+          requires:
+            - build
+            - deploy
+          filters:
+            tags:
+              only: /[0-9]+(\.[0-9]+)*/
+            branches:
+              ignore: /.*/
 jobs:
   build:
     docker:
-    - image: circleci/python:3.6
+      - image: circleci/python:3.6
     steps:
-    - checkout
-    - run:
-        name: Setup testing environment
-        command: |
-          make install-dev
-    - run:
-        name: Run tests and measure coverage
-        command: |
-          make coverage
-    - store_artifacts:
-        path: htmlcov
-workflows:
-  version: 2
-  workflow:
-    jobs:
-    - build
+      - checkout
+      - restore_cache:
+          key: v1-dependency-cache-{{ checksum "setup.py" }}-{{ checksum "Makefile" }}
+      - run:
+          name: Install Python Dependencies
+          command: |
+            python3 -m venv venv
+            . venv/bin/activate
+            make dev
+      - save_cache:
+          key: v1-dependency-cache-{{ checksum "setup.py" }}-{{ checksum "Makefile" }}
+          paths:
+            - "venv"
+      - run:
+          name: Run Tests
+          command: |
+            . venv/bin/activate
+            make coverage
+      - run:
+          name: Git command testing
+          command: |
+            git status
+      - store_artifacts:
+          path: htmlcov/
+  test-python-install:
+    parameters:
+      version:
+        type: string
+        default: latest
+    docker:
+      - image: circleci/python:<< parameters.version >>
+    steps:
+      - checkout
+      - restore_cache:
+          key: v1-dependency-cache-{{ checksum "setup.py" }}-{{ checksum "Makefile" }}
+      - run:
+          name: Install Python Dependencies
+          command: |
+            python3 -m venv venv
+            . venv/bin/activate
+            make dev
+      - save_cache:
+          key: v1-dependency-cache-{{ checksum "setup.py" }}-{{ checksum "Makefile" }}
+          paths:
+            - "venv"
+      - run:
+          name: Run Tests
+          command: |
+            . venv/bin/activate
+            make coverage
+      - store_artifacts:
+          path: htmlcov/
+  deploy:
+    docker:
+      - image: circleci/python:3.6
+    steps:
+      - checkout
+      - restore_cache:
+          key: v1-dependency-cache-{{ checksum "setup.py" }}-{{ checksum "Makefile" }}
+      - run:
+          name: Install Python Dependencies
+          command: |
+            python3 -m venv venv
+            . venv/bin/activate
+            make dev
+      - save_cache:
+          key: v1-dependency-cache-{{ checksum "setup.py" }}-{{ checksum "Makefile" }}
+          paths:
+            - "venv"
+      - run:
+          name: verify git tag vs. version
+          command: |
+            python3 -m venv venv
+            . venv/bin/activate
+            python setup.py verify
+      - run:
+          name: init .pypirc
+          command: |
+            echo -e "[pypi]" >> ~/.pypirc
```

Comment:
Could you put this into the Makefile? It makes the codebase more portable/less dependent on circleci.

Reply:
This goes for the other bits of code in this file.

Reply:
Can you elaborate on what code you are referring to?  The section you specifically tagged is just dropping a resource file into the environment to then call as a reference when doing the upload via twine.

Reply:
I literally mean

```
            echo -e "[pypi]" >> ~/.pypirc
            echo -e "repository =  https://pypi.org/legacy/" >> ~/.pypirc
            echo -e "username = pythonkol" >> ~/.pypirc
            echo -e "password = $PYPI_PASSWORD" >> ~/.pypirc
            echo -e
            echo -e "[testpypi]" >> ~/.pypirc
            echo -e "repository =  https://test.pypi.org/legacy/" >> ~/.pypirc
            echo -e "username = pythonkol" >> ~/.pypirc
            echo -e "password = $PYPI_PASSWORD" >> ~/.pypirc
```

We could instead do this with `make build-pypirc $PYPI_USERNAME $PYPI_PASSWORD`

Reply:
Two things with doing it that way are of concern to me:

- The variables are only present in the CI environment variables and aren't present anywhere else so moving it to make would require someone to know they are required.
- Exposing the variables via the command line would possibly expose it to logging or stdout that would not be ideal.  Since pypi doesn't support token based auth, we'll need to limit exposure of the account passwords as much as possible.

Calling this code is a little bit of stretch.  It's more akin to setting environment variables via export, except this process doesn't support that (would be nice though because then we could use the variables we've already set and have it use them).